src/prefect/orion/orchestration/rules.py

```diff
@@ -268,18 +250,32 @@ async def _validate_proposed_state(
     ):
         if self.proposed_state is None:
             validated_orm_state = self.run.state
+            state_data = None
         else:
+            state_payload = self.proposed_state.dict(shallow=True)
+            state_data = state_payload.pop("data", None)
+
+            if state_data is not None:
+                state_result_artifact = core.Artifact(artifact_data=state_data, flow_run_id=self.run.id)
+                created_artifact = await artifacts.create_artifact(
+                    self.session, state_result_artifact
+                )
+                state_payload["result_artifact_id"] = state_result_artifact.id
+
             validated_orm_state = db.FlowRunState(
                 flow_run_id=self.run.id,
-                **self.proposed_state.dict(shallow=True),
+                **state_payload,
             )
-            self.session.add(validated_orm_state)
-            self.run.set_state(validated_orm_state)
+
+        self.session.add(validated_orm_state)
+        self.run.set_state(validated_orm_state)
 
         await self.session.flush()
-        self.validated_state = (
+        validated_state = (
             validated_orm_state.as_state() if validated_orm_state else None
         )
+        validated_state.data = state_data
```

Comment:
Won't this fail when the `validated_state` is `None`?

Reply:
you're right, I think it will but I also thought `validated_state` will never be `None` anymore except for `None -> None` transitions?

Reply:
It can be set to `None` in the line just above though so we need to handle it